Add marks

kirillkormilcev opened the pull request.

kirillkormilcev committed:
RATE.LIKES in schema
likeFilm() in FilmController
addLike() in LikeService
addLike() in LikeStorage
Film, new Film

kirillkormilcev committed:
RATE.LIKES in schema
likeFilm() in FilmController
addLike() in LikeService
addLike() in LikeStorage
Film, new Film

kirillkormilcev committed:
#add-marks

kirillkormilcev committed:
RATE.LIKES in schema
likeFilm() in FilmController
addLike() in LikeService
addLike() in LikeStorage
Film, new Film
rating validation

kirillkormilcev committed:
rating validation

kirillkormilcev committed:
rating validation

kirillkormilcev committed:
Merge remote-tracking branch 'origin/add-marks' into add-marks

kirillkormilcev committed:
Merge remote-tracking branch 'origin/add-marks' into add-marks

kirillkormilcev committed:
FILMS.RATING

kirillkormilcev committed:
fix texts bugs

kirillkormilcev committed:
remove drop tables from schema
precision

kirillkormilcev committed:
all sorts by rating
update recommendation method

kirillkormilcev committed:
fix recommendation method
warnings

kirillkormilcev committed:
fix pom

kirillkormilcev committed:
Merge remote-tracking branch 'origin/add-marks' into add-marks

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/LikeService.java:
```diff
@@ -2,37 +2,28 @@
-        likeStorage.addLike(filmID, userID);
-        userStorage.addEvent(userID,filmID,"LIKE","ADD");
+    public void addLike(Integer filmID, Integer userID, Integer rate) {
+        if (rate < 1 | rate > 10) {
```
Можем проверить аннотациями в контроллере

GGk77 replied:
Добавил в контроллер проверку аннотациями, "@PositiveOrZero ""@Max(10)"
"Min(0)""@Max(10)" идея не разрешает почему-то

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -89,11 +89,11 @@ public List<Film> getSortedFilms(Integer directorId, String sort) {
     public List<Film> getMostPopularFilms(int genreId, int year, int limit) {
```
параметры genreId и year - опциональные.
ожидается реализация, когда их тип Integer 
так же в контроллере

GGk77 replied:
Исправил на Integer  в контроллере и сервисе

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/LikeService.java:
```diff
@@ -2,37 +2,28 @@
+            throw new ValidationException("Оценка фильма должна быть в диапазоне от 1 до 10 включительно.");
+        }
+        likeStorage.addLike(filmID, userID, rate);
+        likeStorage.updateFilmRatingById(filmID);
```
вызов updateFilmRatingById(filmID) перенести addLike
сервис не знает про хранение, этот ответственность сториджа

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/LikeService.java:
```diff
@@ -2,37 +2,28 @@
-        List<Film> films = likeStorage.getLikedUsersID(count);
-        FilmParameterStorage.loadFilmParameters(films);
-        return films;
+        likeStorage.updateFilmRatingById(filmID);
```
вызов updateFilmRatingById(filmID) перенести deleteLike
сервис не знает про хранение, этот ответственность сториджа

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationService.java:
```diff
@@ -0,0 +1,43 @@
+            recommendedFilms.add(filmStorage.getFilmByID(filmId).get());
+        }
+        filmParameterStorage.loadFilmParameters(recommendedFilms);
+        Comparator<Film> sortByRate = (o1, o2) -> {
```
Сортировку перенести в SQL в вызов findRecommendedFilmIds(userId)

kirillkormilcev replied:
взял в работу

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationService.java:
```diff
@@ -0,0 +1,43 @@
+        }
+        List<Film> recommendedFilms = new ArrayList<>();
+        for (Integer filmId : recommendedFilmIds) {
+            recommendedFilms.add(filmStorage.getFilmByID(filmId).get());
```
Загрузить сразу все фильмы по recommendedFilmIds
Одним запросом вместо N

kirillkormilcev replied:
взял в работу

IvanKlimkin replied:
Предлагаю использовать список как параметр в запросе sql, как в методе loadFilmParameters в классе FilmParameterStorage, что бы получить список фильмов по списку Id.

kirillkormilcev replied:
Спасибо, это похоже то, что нужно, как раз дошел до этого

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/DirectorStorage.java:
```diff
@@ -34,7 +35,7 @@ public Director addDirector(Director director) {
+        director.setId(Objects.requireNonNull(keyHolder.getKey()).intValue());
         log.debug("Сохранен режиссер {}", director.getName());
         return director;
     }
```
update DIRECTORS set DIRECTOR_ID = ?,DIRECTOR_NAME = ? where DIRECTOR_ID = ?
тут лишнее в set
DIRECTOR_ID = ?,

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -36,13 +34,14 @@ public static Film makeFilm(ResultSet rs) throws SQLException {
                 new Mpa(rs.getInt("MPA_ID"), rs.getString("MPA_NAME")),
                 new HashSet<>(),
-                new HashSet<>()
+                new HashSet<>(),
```
добавить дополнительный конструктор, где поле поля специализируется  new HashSet<>()

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -51,12 +50,12 @@ public List<Film> getSortedFilms(Director director, String sort) {
         String sql;
         if (sort.equals("year")) {
```
Лучше реализация есть типы сортировка это enum
String sort -> Order sort
где Order enum

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -51,12 +50,12 @@ public List<Film> getSortedFilms(Director director, String sort) {
                     " left join LIKES L on FILMS.FILM_ID = L.FILM_ID " +
                     "join FILM_DIRECTOR FD on FILMS.FILM_ID = FD.FILM_ID" +
-                    " where FD.DIRECTOR_ID=? group by FILMS.FILM_ID order by COUNT(L.USER_LIKED_ID)";
+                    " where FD.DIRECTOR_ID=? group by FILMS.FILM_ID ORDER BY (AVG(RATING), COUNT(L.USER_LIKED_ID)) DESC";
```
ожидалась та же сортировка что и в предыдущем запросе по rating

vlad-iv replied:
ORDER BY RATING DESC

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
Готово, только пришлось добавить avg, т.к. записи группируются, возможно можно по другому, но не знаю как.

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -71,11 +70,12 @@ public Optional<Film> getFilmByID(Integer ID) {
                     filmRows.getString("NAME"),
                     filmRows.getString("DESCRIPTION"),
-                    filmRows.getDate("RELEASE_DATE").toLocalDate(),
+                    Objects.requireNonNull(filmRows.getDate("RELEASE_DATE")).toLocalDate(),
```
это ответственность бд через NOT NULL в схеме
requireNonNull не нужно

GGk77 replied:
поправил

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -136,42 +136,42 @@ public List<Film> searchFilms(String lowerQuery, String params) {
         String sql;
         if (params.contains("title") && params.contains("director")) {
```
params лучше через enum FilmSearchBy

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -136,42 +136,42 @@ public List<Film> searchFilms(String lowerQuery, String params) {
+                    "from LIKES group by FILM_ID) e on f.FILM_ID = e.FILM_ID " +
+                    "where (lower(d.DIRECTOR_NAME) like '%' || lower(?) || '%' " +
+                    "or lower(f.NAME) like '%' || lower(?) || '%') " +
+                    "ORDER BY RATING, e.evaluate DESC";
```
e.evaluate уже лишнее  везде убираем включая лишние джойны
есть поле RATING

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
Готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -183,7 +183,7 @@ public List<Film> getMostPopularFilmsByYear(int year, int limit) {
                 "where YEAR(F.RELEASE_DATE) = ? " +
                 "group by F.FILM_ID, L.USER_LIKED_ID " +
-                "order by COUNT(L.USER_LIKED_ID) desc limit ?";
+                "ORDER BY (AVG(RATING), COUNT(L.USER_LIKED_ID)) DESC limit ?";
```
COUNT(L.USER_LIKED_ID) везде убираем, AVG тоже и join LIKES не нужен
только RATING

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeStorage.java:
```diff
@@ -27,10 +35,10 @@ public void deleteLike(Integer filmID, Integer userID) {
+        String sql = "select F.*, MPA_NAME from FILMS as F left join LIKES as L on F.FILM_ID=L.FILM_ID " +
                 "JOIN MPA M on M.MPA_ID = F.MPA_ID" +
-                " group by F.FILM_ID  order by COUNT(L.USER_LIKED_ID) DESC LIMIT ?";
+                " group by F.FILM_ID ORDER BY (AVG(RATING), COUNT(L.USER_LIKED_ID)) DESC LIMIT ?"; //, COUNT(L.USER_LIKED_ID)
```
поправить ORDER BY

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeStorage.java:
```diff
@@ -46,14 +54,72 @@ public JdbcTemplate getJdbcTemplate() {
         return function.apply(parameter);
     }
+
+    public void updateFilmRatingById(int filmId) {
```
метод private 
добавить вызов в addLike и deleteLike

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeStorage.java:
```diff
@@ -46,14 +54,72 @@ public JdbcTemplate getJdbcTemplate() {
+        try {
+            for (Rate rate : likedFilmIdsAndRate) {
+                String sql = "select L.USER_LIKED_ID from LIKES L " +
+                        "where (L.FILM_ID = ? and L.RATE >= (? - 2) and L.RATE <= (? + 2))";
```
если поставить like с рейтом 6
то он становится равным дизлайку (like = 4 и 5)
поправить логику что учитываем с RATE > 5:
1-5 понимаем как dislike
6-10 как like

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
готово

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeStorage.java:
```diff
@@ -46,14 +54,72 @@ public JdbcTemplate getJdbcTemplate() {
+                        rate.getRate())); //
+            }
+            userIdsWithSameRates.remove((Integer) userId); //убрать собственный id
+        } catch (EmptyResultDataAccessException e) { // если иных оценок нет или совпадает с оценкой пользователя
```
для чего обработка EmptyResultDataAccessException  
если он выпадает, может другой вызов query 
подобрать верный вызов вместо jdbcTemplate.queryForList и убрать обработку ошибки

kirillkormilcev replied:
взял в работу

kirillkormilcev replied:
Поправил, падало с queryForObject, с queryForList норм, забыл убрать try catch

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/LikeStorage.java:
```diff
@@ -46,14 +54,72 @@ public JdbcTemplate getJdbcTemplate() {
+        }
+        List<Integer> filmIdsRatedByUsersWithSameTastesButNotLikedByUser = new ArrayList<>();
+        for (Integer userIdWithSomeTastes : userIdsWithSameRates) {
+            List<Rate> likedFilmIdsAndRateByUserWithSameTastes = getRatesByUserID(userIdWithSomeTastes);
```
заменить N запросов  getRatesByUserID на 1
лучше отредактировать и больше обработки перенести на SQL

kirillkormilcev replied:
взял в работу

vlad-iv commented on src/main/java/ru/yandex/practicum/filmorate/storage/UserDbStorage.java:
```diff
@@ -60,7 +61,7 @@ public Optional<User> getUserByID(Integer ID) {
                     userRows.getString("EMAIL"),
                     userRows.getString("LOGIN"),
-                    userRows.getDate("BIRTHDAY").toLocalDate(),
+                    Objects.requireNonNull(userRows.getDate("BIRTHDAY")).toLocalDate(),
```
это ответственность бд
requireNonNull не нужен

GGk77 replied:
Исправил

kirillkormilcev committed:
fix: remove try-catch from findRecommendedFilmIds

kirillkormilcev committed:
Film +countPositive, +countNegative
FILMS.COUNT_POSITIVE, COUNT_NEGATIVE
use of updateFilmRatingById to DAO
LIKES.POSITIVITY
refactor updateFilmRatingById to updateFilmRatingAndCountsById

kirillkormilcev commented:
@vlad-iv  По сортировкам: помимо сортировки по рейтингу (во всех выдачах изменено) во вторую и третью очередь добавлена сортировка по количеству позитивных оценок и количеству негативных оценок. Т.е., если фильмы имеют равный рейтинг первым будет выдан тот у кого больше количество положительных оценок и меньше - негативных.
Это не противоречит ТЗ и замечаниям.
Что это дает:
- проходят старые тесты, где важно количество лайков
- обратная совместимость с системой оценок лайками
- делает систему рейтинга более совершенной
- количество лайков как таковое исключается из модели
Хотим представить это как конкурентное преимущество в продукте нашей команды))

kirillkormilcev committed:
sort by rating,countPositive, countNegative in all methods
optimise getSharedFilmsWithFriend()

kirillkormilcev committed:
recommendations RATE > 5
optimise load genres and directors to film

kirillkormilcev committed:
fix

GGk77 committed:
fix: refactoring after review Vladimir
